Jailhouse (inmate/guest) port

glima opened the pull request:
This is code that brings in configuration to build, on x86, a target that will run as a Jailhouse (https://github.com/siemens/jailhouse) hypervisor guest OS. This is sure to work on QEMU, but should work as well on bare metal (provided we have port IO based UART there for console output). It should be tested against their "next" branch, instead of the master one (the cadence of flushes to master is not high, but should happen soon). The licensing of the small code taken from them should match ours.

glima commented:
Fixed last commit's one-char-too-long commit message line.

glima commented:
One more 73 char long line fixed (I wish they were all reported at once).

glima commented:
```
$ ./scripts/sanitycheck -s tests/drivers/pci_enum/test
Cleaning output directory /home/glchaves/devel/iot/zephyr/zephyr/sanity-out
Selecting default platforms per test case
Building testcase defconfigs...
1 tests selected, 27449 tests discarded due to filters
total complete:    1/   1  100%  failed:    0
1 of 1 tests passed with 0 warnings in 5 seconds
```
is my local run of the test that apparently fails, I have no idea what the CI is doing differently.
@nashif , care to give a 1st look, when possible? :)

nashif commented:
will take a look.

nashif commented:
@andrewboie  can you please take a look at this?

andrewboie commented on arch/x86/core/crt0.S:
```diff
@@ -15,6 +15,93 @@
 #include <kernel_arch_data.h>
 #include <arch/cpu.h>
 
+#ifdef CONFIG_JAILHOUSE
```
Can you move this below all the GDATA/GTEXT declarations?

glima replied:
OK

andrewboie commented on arch/x86/core/crt0.S:
```diff
@@ -15,6 +15,93 @@
+	rep stosl
+
+	mov $__start,%ebx
+	mov $_interrupt_stack,%esp
```
__start sets up and switches to the interrupt stack, don't think you need to do this

andrewboie commented on arch/x86/core/crt0.S:
```diff
@@ -15,6 +15,93 @@
+	mov %eax,%es
+	mov %eax,%ss
+
+	xor %eax,%eax
```
why duplicate bss zero logic? __start also does it

glima replied:
Indeed, thanks.

andrewboie commented on arch/x86/core/crt0.S:
```diff
@@ -15,6 +15,93 @@
+	mov $loader_pdpt,%eax
+	mov %eax,%cr3
+
+	/* Enable paging, write protect and protected mode */
```
I'd like to learn more about why we are enabling paging here. Jailhouse requirement?

glima replied:
I guess they wanted that completed at this boot phase there, I see we don't need this here. Thanks.

andrewboie commented on arch/x86/core/crt0.S:
```diff
@@ -15,6 +15,93 @@
+
+	mov $__start,%ebx
+	mov $_interrupt_stack,%esp
+	call *%ebx
```
why a 'call' instead of an unconditional jump?
actually it looks like you can just fall through to __start

glima replied:
Unconditional jump it is, now. Thanks.

andrewboie commented on arch/x86/core/msr.c:
```diff
@@ -75,3 +75,18 @@ u64_t _MsrRead(unsigned int msr)
+#ifdef CONFIG_JAILHOUSE_X2APIC
+#define MSR_X2APIC_BASE 0x00000800
+
+u32_t read_x2apic(unsigned int reg)
```
could maybe make these inline functions in kernel_arch_func.h

glima replied:
okay

andrewboie commented on boards/x86/qemu_x86/qemu_x86_jailhouse_defconfig:
```diff
@@ -0,0 +1,21 @@
+CONFIG_TEST_RANDOM_GENERATOR=y
+CONFIG_XIP=n
+CONFIG_X86_MMU=y
+CONFIG_HW_STACK_PROTECTION=n
```
any reason why we're disabling stack protection?

glima replied:
I have to figure out why `mmu_tables` section wants to end up at an address we have to use here:
`section boot LMA [000000000000fff0,000000000000ffff] overlaps section mmu_tables LMA [000000000000f1f4,0000000000013fff]`. Can this come later? Any clues?

andrewboie replied:
i'd really rather this get root-caused. the problem is likely in the x86 linker script.

glima replied:
What the patch does WRT boot section is tricky too: it injects that boot section with that hardcoded LMA (that is Jailhouse's entry point, we can't escape that) and the rest of the linker script is not actually well informed of that. Now wondering how to proceed better. Maybe shift LMA of all data sections to be after that?

glima replied:
I've been working on this on partial time, but here goes. I've reached people at freenode/##binutils and even they confirm that there is no way/command to tell the linker: "skip this region when assigning LMAs, to account for a hole". We could have 3 slices at ia32/linker.ld (low, boot, high), but again the linker has no support for sections auto-packing, with redirections for more than one memory region via the > operator. We could, finally, have only 2 memory regions there (ram_low, ram) to try and have as few sections as possible patched (only up to the text one in the lower one), but the linker is messing big time with VMAs, still trying to use some that were in the first mem. region in the other. This is very frustrating :/

glima replied:
Solved after interaction with Jailhouse upstream. Now for the last issue.

andrewboie commented on drivers/interrupt_controller/loapic_intr.c:
```diff
@@ -205,23 +206,44 @@ static int _loapic_init(struct device *unused)
 	s32_t loApicMaxLvt; /* local APIC Max LVT */
 
 	/* enable the Local APIC */
+#ifndef CONFIG_JAILHOUSE_X2APIC
```
lots of ifdefs...could we perhaps make an inline function which does the right thing based on configuration?

/* OR in a bit to the existing value*/
LOAPIC_WRITE(LOAPIC_SVR, LOAPIC_READ(LOAPIC_SVR) | LOAPIC_ENABLE);

glima replied:
Ack.

glima replied:
loapic_timer.c will be left as is, though, because I see no advantages there.

andrewboie commented on include/arch/x86/linker.ld:
```diff
@@ -180,10 +205,18 @@ SECTIONS
+	 * Because of text section displacement to follow startup ones
+	 * on LMA, align back to page here
+	 */
+#ifdef CONFIG_JAILHOUSE
```
I'm confused by this. Can you provide more detail?

glima replied:
Maybe I have to rephrase the comment, help is appreciated. With this code, I get

```
In Page directory entry 0    : valid start address = 0x000f4000, end address = 0x003fffff
In Page directory entry 0    : valid start address = 0x00000000, end address = 0x000f3fff
```
as PDE entries for those ranges, without I get

`Memory Regions are not page aligned 0xf3c40`

instead.

andrewboie replied:
somehow we are passing an unaligned address/size to an MMU_BOOT_REGION() call. What MMU_BOOT_REGION call is generating this error? What about this patch shifts it?

glima replied:
It seems to be `__kernel_ram_start`. I have no clue why originally it would end up page aligned by just reading that linker code. What we shift on purpose is the LMA of the text section, to be placed just after the (new) startup one. This is how Jailhouse works, by having the jump from the 16-bit (boot) section to the proper 32-bit text be a fall-through. For now I'll keep looking, any help is appreciated.

glima replied:
It seems the page alignment comes for free from `PHYS_RAM_ADDR`, since it's the first section in that memory region. With this port, we lay out every section in a sequence and we may end up unaligned. Either we ease on the checking script or we keep this line. I'll pack all the patches up again as soon as I work with upstream Jailhouse to have the new, necessary patches on their "next" branch.

andrewboie commented on Makefile:
```diff
@@ -841,14 +841,21 @@ libzephyr.a: $(filter-out %/lib.a, $(core-y))
 
+K_ENTRY=''
+ifeq ($(CONFIG_JAILHOUSE),y)
+	K_ENTRY='__jh_entry'
```
hrm. I think we should perhaps make the starting symbol a config option

glima replied:
Will do.

Commit:
build: Only have MMU_BOOT_REGION decl. for HPET if the latter is on

This would impact builds with no CONFIG_HPET_TIMER=y set.

Signed-off-by: Gustavo Lima Chaves <gustavo.lima.chaves@intel.com>

Commit:
build: make kernel entry symbol a config option

This might change for different ports, so make it configurable.

Signed-off-by: Gustavo Lima Chaves <gustavo.lima.chaves@intel.com>

Commit:
loapic_intr: factor (out) APIC accesses in static functions

This will accomodate newer access models later, with variations of those
functions' contents.

Signed-off-by: Gustavo Lima Chaves <gustavo.lima.chaves@intel.com>

Commit:
x86: Jailhouse port, tested for UART (# 0, polling) and LOAPIC timer

This is an introductory port for Zephyr to be run as a Jailhouse
hypervisor[1]'s "inmate cell", on x86 64-bit CPUs (running on 32-bit
mode). This was tested with their "tiny-demo" inmate demo cell
configuration, which takes one of the CPUs of the QEMU-VM root cell
config, along with some RAM and serial controller access (it will even
do nice things like reserving some L3 cache for it via Intel CAT) and
Zephyr samples:

   - hello_world
   - philosophers
   - synchronization

The final binary receives an additional boot sequence preamble that
conforms to Jailhouse's expectations (starts at 0x0 in real mode). It
will put the processor in 32-bit protected mode and then proceed to
Zephyr's __start function.

Testing it is just a matter of:
  $ mmake -C samples/<sample_dir> BOARD=x86_jailhouse JAILHOUSE_QEMU_IMG_FILE=<path_to_image.qcow2> run
  $ sudo insmod <path to jailhouse.ko>
  $ sudo jailhouse enable <path to configs/qemu-x86.cell>
  $ sudo jailhouse cell create <path to configs/tiny-demo.cell>
  $ sudo mount -t 9p -o trans/virtio host /mnt
  $ sudo jailhouse cell load tiny-demo /mnt/zephyr.bin
  $ sudo jailhouse cell start tiny-demo
  $ sudo jailhouse cell destroy tiny-demo
  $ sudo jailhouse disable
  $ sudo rmmod jailhouse

For the hello_world demo case, one should then get QEMU's serial port
output similar to:

"""
Created cell "tiny-demo"
Page pool usage after cell creation: mem 275/1480, remap 65607/131072
Cell "tiny-demo" can be loaded
CPU 3 received SIPI, vector 100
Started cell "tiny-demo"
***** BOOTING ZEPHYR OS v1.9.0 - BUILD: Sep 12 2017 20:03:22 *****
Hello World! x86
"""

Note that the Jailhouse's root cell *has to be started in xAPIC
mode* (kernel command line argument 'nox2apic') in order for this to
work. x2APIC support and its reasoning will come on a separate commit.

As a reminder, the make run target introduced for x86_jailhouse board
involves a root cell image with Jailhouse in it, to be launched and then
partitioned (with >= 2 64-bit CPUs in it).

Inmate cell configs with no JAILHOUSE_CELL_PASSIVE_COMMREG flag
set (e.g. apic-demo one) would need extra code in Zephyr to deal with
cell shutdown command responses from the hypervisor.

You may want to fine tune CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC for your
specific CPU—there is no detection from Zephyr with regard to that.

Other config differences from pristine QEMU defaults worth of mention
are:

   - there is no HPET when running as Jailhouse guest. We use the LOAPIC
     timer, instead
   - there is no PIC_DISABLE, because there is no 8259A PIC when running
     as a Jailhouse guest
   - XIP makes no sense also when running as Jailhouse guest, and both
     PHYS_RAM_ADDR/PHYS_LOAD_ADD are set to zero, what tiny-demo cell
     config is set to

This opens up new possibilities for Zephyr, so that usages beyond just
MCUs come to the table. I see special demand coming from
functional-safety related use cases on industry, automotive, etc.

[1] https://github.com/siemens/jailhouse

Reference to Jailhouse's booting preamble code:

Origin: Jailhouse
License: BSD 2-Clause
URL: https://github.com/siemens/jailhouse
commit: 607251b44397666a3cbbf859d784dccf20aba016
Purpose: Dual-licensing of inmate lib code
Maintained-by: Zephyr

Signed-off-by: Gustavo Lima Chaves <gustavo.lima.chaves@intel.com>

Commit:
jailhouse: add support for x2APIC mode for all LOAPIC accesses

Besides the fact that we did not have that for the current supported
boards, that makes sense for this new, virtualized mode, that is meant
to be run on top of full-fledged x86 64 CPUs.

By having xAPIC mode access only, Jailhouse has to intercept those MMIO
reads and writes, in order to examine what they do and arbitrate if it's
safe or not (e.g. not all values are accepted to ICR register). This
means that we can't run away from having a VM-exit event for each and
every access to APIC memory region and this impacts the latency the
guest OS observes over bare metal a lot.

When in x2APIC mode, Jailhouse does not require VM-exits for MSR
accesses other that writes to the ICR register, so the latency the guest
observes is reduced to almost zero.

Here are some outputs of the the command line

  $ sudo ./tools/jailhouse cell stats tiny-demo

on a Jailhouse's root cell console, for one of the Zephyr demos using
LOAPIC timers, left for a couple of seconds:

Statistics for tiny-demo cell (x2APIC root, x2APIC inmate)

COUNTER                              SUM   PER SEC
vmexits_total                          7         0
vmexits_management                     3         0
vmexits_cr                             2         0
vmexits_cpuid                          1         0
vmexits_msr                            1         0
vmexits_exception                      0         0
vmexits_hypercall                      0         0
vmexits_mmio                           0         0
vmexits_pio                            0         0
vmexits_xapic                          0         0
vmexits_xsetbv                         0         0

Statistics for tiny-demo cell (xAPIC root, xAPIC inmate)

COUNTER                              SUM   PER SEC
vmexits_total                       4087        40
vmexits_xapic                       4080        40
vmexits_management                     3         0
vmexits_cr                             2         0
vmexits_cpuid                          1         0
vmexits_msr                            1         0
vmexits_exception                      0         0
vmexits_hypercall                      0         0
vmexits_mmio                           0         0
vmexits_pio                            0         0
vmexits_xsetbv                         0         0

Statistics for tiny-demo cell (xAPIC root, x2APIC inmate)

COUNTER                              SUM   PER SEC
vmexits_total                       4087        40
vmexits_msr                         4080        40
vmexits_management                     3         0
vmexits_cr                             2         0
vmexits_cpuid                          1         0
vmexits_exception                      0         0
vmexits_hypercall                      0         0
vmexits_mmio                           0         0
vmexits_pio                            0         0
vmexits_xapic                          0         0
vmexits_xsetbv                         0         0

See that under x2APIC mode on both Jailhouse/root-cell and guest, the
interruptions from the hypervisor are minimal. That is not the case when
Jailhouse is on xAPIC mode, though. Note also that, as a plus, x2APIC
accesses on the guest will map to xAPIC MMIO on the hypervisor just
fine.

Signed-off-by: Gustavo Lima Chaves <gustavo.lima.chaves@intel.com>

nashif commented:
recheck

glima commented:
@andrewboie ,

https://github.com/siemens/jailhouse/commits/next has been updated with the required patches for this new series version to work and I think all your comments have been addressed. Can you please take a second look?

glima commented:
CI has some style issues reported, mostly things that were an issue before the refactoring commit, now gone, I hope.

andrewboie commented on arch/x86/core/crt0.S:
```diff
@@ -42,6 +42,84 @@
+	or $X86_CR4_PSE,%eax
+	mov %eax,%cr4
+
+	mov $loader_pdpt,%eax
```
Do we really need to set a page table here? Why not leave un-configured and just let __start deal with it?

glima replied:
You are right, there is no use until there.

andrewboie commented on arch/x86/core/crt0.S:
```diff
@@ -42,6 +42,84 @@
+	mov $(X86_CR0_WP | X86_CR0_PE),%eax
+	mov %eax,%cr0
+
+	movl $MSR_MTRR_DEF_TYPE,%ecx
```
Is this something specific for Jailhouse?
Consider just moving this to __start

glima replied:
I guess so, but correct me otherwise. Do Intel MCUs have this MSR as well? I guess this was put on the boot code to enable caching on simpler guests (the demos on Jailhouse), but I'm still confused because the manual says "When the E flag is set, the type specified in the default memory type field is used for areas of memory not already mapped by either a fixed or variable MTRR", and that type is *uncached*. @jan-kiszka , any input?

jan-kiszka replied:
This is IA32_MTRR_DEF_TYPE on Intel, and it is disabled on reset so that memory becomes uncached. Jailhouse hands the CPU over in reset state, therefore we have to enable it explicitly. On a normal system, the firmware tends to enable MTRRs already.

andrewboie replied:
ok, i'm fine with leaving this where it is

andrewboie commented on arch/x86/core/crt0.S:
```diff
@@ -42,6 +42,84 @@
+	ljmp $JAILHOUSE_INMATE_CS32, $__start
+
+	.global loader_gdt
+loader_gdt:
```
consider defining the GDT here in a more human-readable way, see how gdt_rom is defined at the end of this file, with comments explaining all values.

glima replied:
Fair.

andrewboie commented on boards/x86/qemu_x86/qemu_x86_jailhouse_defconfig:
```diff
@@ -0,0 +1,22 @@
+CONFIG_TEST_RANDOM_GENERATOR=y
+CONFIG_XIP=n
+CONFIG_X86_MMU=y
+CONFIG_HW_STACK_PROTECTION=y
```
add CONFIG_USERSPACE=y and CONFIG_X86_PAE_MODE=y to match qemu_x86

glima replied:
I guess I'll have to move the whole pull request forward to get these. A new pull request, then?

andrewboie replied:
not sure what you mean by this, are you saying your branch is out of date? you should be able to just rebase it on master and force-push it

glima replied:
I was afraid GitHub was not smart enough to get that. Alright, will do.

glima replied:
Now looking at this new DTS config introduction (ddf6a69). How is one suppose to configure that? Only by providing another one from scratch?

glima replied:
DTSes don't expect `#ifdef`s  on them and I don't want to mess with the `CONFIG_BOARD` entry, which influences which file extract_dts_includes.py expects :( I'll come back later.

tyrantlizard77 reviewed (changes requested):
just some nits, also a little concerned that the minimal __jh_entry loader may still be doing a little too much

Commit:
x86: add new board to accomodate Jailhouse port

The port will enable Zephyr to run as a guest OS on x86-64 systems. It
comes with a test on QEMU to validate that, thus this new board
introduction. It's "make run" target will issue QEMU with the same
configuration Jailhouse upstream uses for their confis/qemu-x86.c root
cell configuration:

  Test configuration for QEMU Q35 VM, 1 GB RAM, 4 cores,
  6 MB hypervisor, 60 MB inmates (-4K shared mem device)

This will work provided qemu-system-x86_64 is installed in the system
and a given (qcow2) image with the Jailhouse root cell in it is
provided (any of those will ever ship with Zephyr, it's out of its
scope).

Signed-off-by: Gustavo Lima Chaves <gustavo.lima.chaves@intel.com>

glima commented:
New DTS config issue attacked by having a dedicated board for the port (@nashif was OK with that). I have now to continue on the issue PAE mode introduced. @jan-kiszka , you can now reproduce it by building and running as instructed by the patches.

nashif commented:
Do we really have to have qemu_x86 in the name of the board? we can just call it jailhouse and still run it in qemu if needed.

jan-kiszka commented:
@glima PAE problem reproduced and understood (took a while...): It's an upstream Zephyr bug.

The page table it generates has XD enable on page directory level, and that naturally kills any execution once you enable paging. I suspect the PAE code of Zepyhr has never been tested on a CPU that also supports EFER.NXE.

glima commented:
Thanks, @jan-kiszka , that makes sense! @nashif I'm fine changing the board name, will do that soon.

lpereira commented:
@AdithyaBaglody Could you please take a look here?

andrewboie commented:
The PAE XD bit issue should be fixed now

dbkinder commented on boards/x86/qemu_x86_jailhouse/doc/board.rst:
```diff
@@ -0,0 +1,120 @@
+Overview
+********
+
+The X86 Jailhouse QEMU board configuration is used to emulate the X86
```
your commit message mentions "This is code that brings in configuration to build, on x86, a target that will run as a Jailhouse (https://github.com/siemens/jailhouse) hypervisor guest OS" but there's no mention of this in the document.   Maybe at least add a reference/link to this github site for more information?

dbkinder commented on boards/x86/qemu_x86_jailhouse/doc/board.rst:
```diff
@@ -0,0 +1,120 @@
+
+        $ make -C samples/synchronization BOARD=qemu_x86_jailhouse JAILHOUSE_QEMU_IMG_FILE=path_to_image.qcow2 run
+
+This assumes the user has the binary qemu-system-x86_64 in their
```
Where does the user get this qemu-system-x86_64 binary image?

dbkinder commented on boards/x86/qemu_x86_jailhouse/doc/board.rst:
```diff
@@ -0,0 +1,120 @@
+
+.. code-block:: shell
+
+        $ sudo insmod <path to jailhouse.ko>
```
Where does jailhouse.ko come from? and all these other files (such as qemu-vm.cell) for that matter?

dbkinder reviewed (changes requested):
I'm thinking this needs some more introduction and references to the Jailhouse github repo in the documentation so folks can know how to work with it?

glima commented:
Hi, @dbkinder . I just pushed a would-be final version, but fair, I'll see how to augment the documentation commit. If you have suggestions, please give them!

jan-kiszka commented:
@glima pushed where?

Note that we are also missing a fix to the APIC timer setup, otherwise Zephyr's periodic timer won't work under KVM. A student of us will send a corresponding PR in a couple of days (busy with his thesis ATM).

Another thing I didn't understand yet: How are is the APIC timer calibrated or configured to a known value? And if the TSC should be used as clock, the same questions applies here.

glima commented:
@jan-kiszka Oops, it seems this got merged. It was pushed here, on the PR.
I don't fully get what the error with APIC timers is since most of the samples that work on the Jailhouse port make use of periodic timers and work just fine.

There is no calibration whatsoever, for now, one has to hardcode `CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC` for now and assume frequencies won't change, too. This needs improvements, I know (will look into it eventually).

I also did not forget the docs, will take a look.

jan-kiszka commented:
@glima Zephry programs the APIC timer prior to setting its mode (and prior to registering a handler, well...). If the timer expires before the periodic mode is enabled, it will not expire again. This can happen with short periods and due to interruptions in this virtualized environment. It happens all the time here.

What CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC value are you using, BTW? Also 150000000?

glima commented:
@jan-kiszka yes, that value (I know it's arbitrary, by now).
You're true about that ordering, I just wonder why I don't see those issues here.
